internal/pkg/v2/application/secret.go

```diff
@@ -0,0 +1,51 @@
+//
+// Copyright (C) 2021 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package application
+
+import (
+	"strings"
+
+	"github.com/edgexfoundry/go-mod-bootstrap/v2/bootstrap/container"
+	"github.com/edgexfoundry/go-mod-bootstrap/v2/di"
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/errors"
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/v2/dtos/common"
+)
+
+// AddSecret adds EdgeX Service exclusive secret to the Secret Store
+func AddSecret(dic *di.Container, request common.SecretRequest) errors.EdgeX {
+	path, secret := prepareSecret(dic, request)
+
+	secretProvider := container.SecretProviderFrom(dic.Get)
+	if secretProvider == nil {
+		return errors.NewCommonEdgeX(errors.KindServerError, "secret provider is missing. Make sure it is specified to be used in bootstrap.Run()", nil)
+	}
+
+	if err := secretProvider.StoreSecret(path, secret); err != nil {
+		return errors.NewCommonEdgeX(errors.KindServerError, "adding secret failed", err)
+	}
+	return nil
+}
+
+func prepareSecret(dic *di.Container, request common.SecretRequest) (string, map[string]string) {
+	secretStoreInfo := container.ConfigurationFrom(dic.Get).GetBootstrap().SecretStore
+
+	var secretsKV = make(map[string]string)
+	for _, secret := range request.SecretData {
+		secretsKV[secret.Key] = secret.Value
+	}
+
+	path := strings.TrimSpace(request.Path)
+
+	// add '/' in the full URL path if it's not already at the end of the base path or sub path
+	if !strings.HasSuffix(secretStoreInfo.Path, "/") && !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	} else if strings.HasSuffix(secretStoreInfo.Path, "/") && strings.HasPrefix(path, "/") {
+		// remove extra '/' in the full URL path because secret store's (Vault) APIs don't handle extra '/'.
+		path = path[1:]
+	}
```

Comment:
Very recent change for the value of  secretStoreInfo.Path. It is no longer the full base base. Just the sub path, i.e. "core-data/" . This code is no longer needed. Lower level code handles it.

Reply:
FYI, I verified this code can be removed using App Service version and have removed it in this PR
https://github.com/edgexfoundry/app-functions-sdk-go/pull/870
I will submit same clean-up for Device SDK.

Reply:
Modified. 

1. But I got a 404 error when the path contains the prefix `/` 
```
curl http://0.0.0.0:59860/api/v2/secret -X POST -H "Content-Type: application/json" -d '{"apiVersion":"v2", "path": "/mqtt",  "secretData":[{"key":"username","value":"tester"}]}' | json_pp

{
   "apiVersion" : "v2",
   "message" : "adding secret failed -> Error found on handling secrets from underlying data-store: Received a '404' response from the secret store",
   "statusCode" : 500
}
```

2. And got 400 error when the path contains the suffix `//`
```
% curl http://0.0.0.0:59860/api/v2/secret -X POST -H "Content-Type: application/json" \
  -d '{"apiVersion":"v2", "path": "mqtt//", "secretData":[{"key":"username","value":"tester"}]}' | json_pp

{
   "apiVersion" : "v2",
   "message" : "adding secret failed -> Error found on handling secrets from underlying data-store: Received a '400' response from the secret store",
   "statusCode" : 500
}
```

I thought we can improve the error message to return the actual error
https://github.com/edgexfoundry/go-mod-secrets/blob/master/internal/pkg/vault/secrets.go#L450

Reply:
Make sure you are using `go-mod-secrets v2.0.0-dev.23` & `go-mod-bootstrap v2.0.0-dev.61`

Reply:
Humm, please set logging to DEBUG so it logs the URL it is attempting to use so we have a better understanding of what is happening. Note `//` is never valid. So that error would be expected. Also please try just `mqtt`  as the expected valid values are `/mqtt` & 'mqtt`

Yes, that error message could be improved. ;-)

Reply:
Got it. Thank you.

Reply:
I just verified with my Device SDK PR, that the two valid options are actually `mqtt` and `mqtt/`, which make more sense for specifying a sub-path

Reply:
i.e. I got the sane errors using `/mqtt` and `/mqtt//`, but no errors using `nqtt` or `mqtt/` or even `mqtt/something`

Reply:
> I got the sane errors using /mqtt and /mqtt//, but no errors using nqtt or mqtt/ or even mqtt/something

Yes, I also verified that. The user just needs to use DEBUG log to inspect and resolve the error.

I have no problem now.  The changes made, please help review.